### ISSUE TITLE
why does letsencrypt.sh not allow parallel calling?

### ISSUE THREAD

**Issue by philkunz** [lukas2511/letsencrypt.sh#245]
why does letsencrypt.sh not allow parallel calling?

I'm trying to speed up https://www.npmjs.com/package/cert which uses letsencrypt.sh in the background for now.

Not being able to call it in parallel means that the process is blocking which is usually bad behaviour, at least if requests are made for different domain names.

**Comment by omega8cc** [comment]
This sounds like a duplicate of #228

**Comment by lukas2511** [comment]
letsencrypt.sh mainly doesn't allow parallel calling because it's a lot easier to just stop if another instance in general is running than to check if another instance is running that is trying to modify the same files (account key generation / registration, private key (re)generation, csr generation, storing of new certificates, moving of old files to archive, hook stuff, ..., a lot of locations in the script where files are written or read)

I added an option in bd9cc5b to disable locks completely, be careful when using this.

**Comment by seefood** [comment]
you can lock per domain and then it's easy to fork requests.
One thing this will solve - I had one line of the `domains.txt` fail, and then the rest don't continue. It's a major problem if you find it in an exploding inbox too late.

**Comment by philkunz** [comment]
@lukas2511  Thanks so much! As far as I am concerned this is perfect now. I can call it and map myself which child processes currently request which domain to avoid read/write problems + I can store logs domain related. I'll add a link to this repo in the description of cert.
